cmd/osbuild-rcm-tests/main_test.go

```diff
@@ -41,40 +41,18 @@ func main() {
 	// Case 1: POST request
 
 	resp, err := http.Post(socket+endpoint, "application/json", strings.NewReader(submitBody))
-	if err != nil {
-		log.Fatal("Failed to submit a compose: ", err.Error())
-	}
-	if resp.StatusCode != 200 {
-		log.Print("Error: the ", endpoint, " returned non 200 status. Full response: ", resp)
-		failed = true
-	} else {
-		err = json.NewDecoder(resp.Body).Decode(&submitResponse)
-		if err != nil {
-			log.Fatal("Failed to decode JSON response from ", endpoint)
-		}
-		log.Print("Success: the ", endpoint, " returned compose UUID: ", submitResponse.UUID)
-	}
+	require.Nilf(t, err, "Failed to submit a compose: %v", err)
+	assert.Equalf(t, resp.StatusCode, 200, "Error: the %v returned non 200 status. Full response: %v", endpoint, resp)
+	err = json.NewDecoder(resp.Body).Decode(&submitResponse)
+	require.Nilf(t, err, "Failed to decode JSON response from %v", endpoint)
```

Comment:
This implementation will attempt to decode the body even in case the HTTP return code wasn't 200. Isn't not a big problem, but I think we could preserve the if statement here so that we don't inform about HTTP failure and JSON failure at the same time. If we preserve it, we can also inform about the success.

Reply:
Shouldn't the body always be valid JSON, even on error codes? If so, I think this is correct, but might warrant a comment.

Reply:
I'm not sure, the body can be anything depending on the content-type, can it? I'd expect it to be empty when the code is 400: https://play.golang.org/p/LcPMoym15-7

Reply:
I think it will print the error, check this if I wrote a correct reproducer:
```
~/t/gohttp [1] $ cat server.go

package main

import (
    "log"
    "net/http"
)

func handler(w http.ResponseWriter, r *http.Request) {
    w.WriteHeader(http.StatusBadRequest)
}

func main() {
    http.HandleFunc("/", handler)
    log.Fatal(http.ListenAndServe(":8080", nil))
}
~/t/gohttp $ cat client.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
)

func main() {
	resp, err := http.Get("http://127.0.0.1:8080/")
	fmt.Println("err: ", err)
	fmt.Println("resp: ", resp)
	fmt.Println("resp.Body: ", resp.Body)
	var statusResponse struct {
		Status string `json:"status"`
	}
	err = json.NewDecoder(resp.Body).Decode(&statusResponse)
	fmt.Println("err: ", err)
	if err != nil {
		fmt.Println("err != nil")
	}
	fmt.Println("statusResponse: ", statusResponse)
	fmt.Println("statusResponse.Status: ", statusResponse.Status)
}                                                                                                                  
~/t/gohttp $ go run server.go &
~/t/gohttp $ go run client.go
err:  <nil>
resp:  &{400 Bad Request 400 HTTP/1.1 1 1 map[Content-Length:[0] Date:[Tue, 17 Mar 2020 08:55:49 GMT]] {} 0 [] false false map[] 0xc000114000 <nil>}
resp.Body:  {}
err:  EOF
err != nil
statusResponse:  {}
statusResponse.Status:
```

Reply:
> I'm not sure, the body can be anything depending on the content-type, can it? I'd expect it to be empty when the code is 400: https://play.golang.org/p/LcPMoym15-7

Right, I mean our API should always return valid JSON, no?

Reply:
I was not sure how to write the asserts and use the if else statement together. Would this make sense? 
```
        if resp.StatusCode != 200 {
￼		require.Nilf(t, err, "Failed to submit a compose: %v", err)
￼	} else {
￼		err = json.NewDecoder(resp.Body).Decode(&submitResponse)
￼	        require.Nilf(t, err, "Failed to decode JSON response from %v", endpoint)
￼		t.Log("Success: the ", endpoint, " returned compose UUID: ", submitResponse.UUID)
￼	}
```

Reply:
Yeah this would work, but I think we always want to treat the reply as JSON, as you had before. Unless I'm misunderstanding something. Let's wait for @msehnout's reply.

Reply:
Yes, we don't return anything else than JSON.

Reply:
`assert.Equalf(t, resp.StatusCode, 200,` -> `require.Equal` and that will short-circuit the test execution not trying to decode JSON in case of HTTP errors.  Wouldn't that be sufficient ?

Reply:
> `assert.Equalf(t, resp.StatusCode, 200,` -> `require.Equal` and that will short-circuit the test execution not trying to decode JSON in case of HTTP errors. Wouldn't that be sufficient ?

Ah yes, I din't think of that, thanks.